app/services/shop_products_service.rb

```diff
@@ -0,0 +1,76 @@
+# Returns a (paginatable) AR object for the products or variants in stock for a given shop and OC.
+# The stock-checking includes on_demand and stock level overrides from variant_overrides.
+
+class ShopProductsService
+  def initialize(distributor, order_cycle)
+    @distributor = distributor
+    @order_cycle = order_cycle
+  end
+
+  def products_relation
+    Spree::Product.where(id: stocked_products)
```

Comment:
having to fetch the product rows from the ids collected in `stocked_products` seems a waste of resources to me. We can start building the AR from `Spree::Product` and group by `spree_variants.product_id` and we will avoid unnecessary code we'll need to maintain, wasted Ruby GC time, slow Ruby processing on the product ids and instantiating bloated ActiveModel objects plus the network round trip, among others.

I suggest we create an issue to fix it in the next sprint.

Reply:
It's still a single AR object right? We're not making an extra query here.

Reply:
You are saying this triggers a singler query where `stocked_products` becomes a subquery? Have you checked if that is the case?

Reply:
Yes, that's the beauty of it! All the methods here only build AR objects with subqueries. This test passes, for example:

```ruby
    describe "stocked_products" do
      it "returns an AR object" do
        stocked_products = described_class.new(distributor, order_cycle).send(:stocked_products)
        expect(stocked_products).to be_an ActiveRecord::Relation
      end
    end
```

Reply:
awesome!